pkg/collector/corechecks/snmp/gosnmp_value.go

```diff
@@ -0,0 +1,154 @@
+		}
+		value = strValue
+	case gosnmp.ObjectIdentifier:
+		strValue, ok := pduVariable.Value.(string)
+		if !ok {
+			return name, snmpValueType{}, fmt.Errorf("oid %s: ObjectIdentifier should be string type but got %T type: %#v", pduVariable.Name, pduVariable.Value, pduVariable)
+		}
+		value = strings.TrimLeft(strValue, ".")
+	default:
+		return name, snmpValueType{}, fmt.Errorf("oid %s: invalid type: %s", pduVariable.Name, pduVariable.Type.String())
+	}
+	submissionType := getSubmissionType(pduVariable.Type)
+	return name, snmpValueType{submissionType: submissionType, value: value}, nil
+}
+
+func hasNonPrintableByte(bytesValue []byte) bool {
+	hasNonPrintable := false
+	for _, bit := range bytesValue {
+		if bit < 32 || bit > 126 {
+			hasNonPrintable = true
+		}
+	}
+	return hasNonPrintable
+}
+
+func resultToScalarValues(result *gosnmp.SnmpPacket) scalarResultValuesType {
+	returnValues := make(map[string]snmpValueType, len(result.Variables))
+	for _, pduVariable := range result.Variables {
+		if shouldSkip(pduVariable.Type) {
+			continue
+		}
+		name, value, err := getValueFromPDU(pduVariable)
+		if err != nil {
+			log.Debugf("cannot get value for variable `%v` with type `%v` and value `%v`", pduVariable.Name, pduVariable.Type, pduVariable.Value)
+			continue
+		}
+		returnValues[name] = value
+	}
+	return returnValues
+}
+
+// resultToColumnValues builds column values
+// - columnResultValuesType: column values
+// - nextOidsMap: represent the oids that can be used to retrieve following rows/values
+func resultToColumnValues(columnOids []string, snmpPacket *gosnmp.SnmpPacket) (columnResultValuesType, map[string]string) {
+	returnValues := make(columnResultValuesType, len(columnOids))
+	nextOidsMap := make(map[string]string, len(columnOids))
+	maxRowsPerCol := int(math.Ceil(float64(len(snmpPacket.Variables)) / float64(len(columnOids))))
+	for i, pduVariable := range snmpPacket.Variables {
+		if shouldSkip(pduVariable.Type) {
+			continue
+		}
+
+		oid, value, err := getValueFromPDU(pduVariable)
+		if err != nil {
+			log.Debugf("Cannot get value for variable `%v` with type `%v` and value `%v`", pduVariable.Name, pduVariable.Type, pduVariable.Value)
+			continue
+		}
+		// the snmpPacket might contain multiple row values for a single column
+		// and the columnOid can be derived from the index of the PDU variable.
+		columnOid := columnOids[i%len(columnOids)]
+		if _, ok := returnValues[columnOid]; !ok {
+			returnValues[columnOid] = make(map[string]snmpValueType, maxRowsPerCol)
+		}
+
+		prefix := columnOid + "."
+		if strings.HasPrefix(oid, prefix) {
+			returnValues[columnOid][oid[len(prefix):]] = value
+			nextOidsMap[columnOid] = oid
+		} else {
+			// if oid is not prefixed by columnOid, it means it's not part of the column
+			// and we can stop requesting the next row of this column
+			delete(nextOidsMap, columnOid)
```

Comment:
Yes, this is expected. We enter this case if the `oid` that is intended to be the instance OID of the column does not start with the column prefix. This mean that we are "out of bound" and there is no more OID to fetch for this column and we drop it from `nextOidsMap`.

This is the case in `Test_fetchColumnOids` where column `1.1.2` has following response values:
- 1.1.2.1
- 1.1.2.2
- 1.1.3.1

The `1.1.3.1` is "out of bound" and we stop request for `1.1.2` column.

If the response is:
- 1.1.2.1
- 1.1.2.2
- 1.1.2.3

`1.1.2.3` would be part of `nextOidsMap` since we are not yet "out of bound".

Reply:
Add a note that this is expected here: https://github.com/DataDog/datadog-agent/commit/c582e592c970ca35d946a918f04896e97b55d655#diff-5df12d78902bfbf0e68d013ff0ad51dbf7cf5bf9a38156a8b26bee35043308c9R127-R128